src/dbnode/storage/index.go

```diff
@@ -1560,17 +1561,47 @@ func (i *nsIndex) queryWithSpan(
 		state = asyncQueryExecState{
 			exhaustive: true,
 		}
-		deadline = start.Add(timeout)
-		wg       sync.WaitGroup
+		deadline      = start.Add(timeout)
+		wg            sync.WaitGroup
+		totalWaitTime time.Duration
 	)
 
 	// Create a cancellable lifetime and cancel it at end of this method so that
 	// no child async task modifies the result after this method returns.
 	cancellable := xresource.NewCancellableLifetime()
+	// always cancel to cleanup. this might be a no-op if it was already canceled by the background task to interrupt
+	// the workers.
 	defer cancellable.Cancel()
 
-	indexMatchingStartTime := time.Now()
-	var totalWaitTime time.Duration
+	stdCtx, ctxFound := ctx.GoContext()
+	if !ctxFound {
+		stdCtx = stdctx.Background()
+	}
+	select {
+	case <-stdCtx.Done():
+		return false, fmt.Errorf("caller canceled the query before it started")
+	default:
+	}
+
+	if timeout > 0 {
+		// launch a goroutine to interrupt the query workers after the timeout.
+		fnDoneCh := make(chan struct{}, 1) // don't block the defer in the case of a timeout.
+		defer func() {
+			fnDoneCh <- struct{}{}
```

Comment:
this is why we really need to start embracing the std go ctx instead. ideally this would be

```
ctx, cancel := WithCancel(ctx)
defer cancel()

select {
   <-ctx.Done()
}
```

however, the m3 `Context` makes that difficult to achieve, since it's doesn't have the same immutability pattern. If you do `Context.SetGoContext(ctx)` you'll impact the ctx of the caller with the new cancel, which is not good.

Reply:
Yeah maybe we should honestly just remove the whole concept of "not having a timeout" here and rely on the context having an existing one?

As @nbroyles discovered it's basically guaranteed you'll have a serverside Go context with a timeout specified already with TChannel Thrift.